src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs

```diff
@@ -1344,40 +1344,57 @@ private void WriteMessageAligned(string message, bool prefixAlreadyWritten, int
+                    string[] nonNullMessages = SplitStringOnNewLines(message);
+                    for (int i = 0; i < nonNullMessages.Length; i++)
                     {
-                        // Our message may have embedded tab characters, so expand those to their space
-                        // equivalent so that wrapping works as expected.
-                        nonNullMessage = nonNullMessage.Replace("\t", consoleTab);
-
-                        // If the message and the prefix are too large for one line in the console, split the string to fit
-                        int index = 0;
-                        int messageLength = nonNullMessage.Length;
-                        // Loop until all the string has been sent to the console
-                        while (index < messageLength)
+                        string nonNullMessage = nonNullMessages[i];
+                        // Take into account the new line char which will be added to the end or each reformatted string
+                        int bufferWidthMinusNewLine = _bufferWidth - 1;
+
+                        // If the buffer is larger then the prefix information (timestamp and key) then reformat the messages. 
+                        // If there is not enough room just print the message out and let the console do the formatting
+                        bool bufferIsLargerThanPrefix = bufferWidthMinusNewLine > adjustedPrefixWidth;
+                        bool messageAndPrefixTooLargeForBuffer = (nonNullMessage.Length + adjustedPrefixWidth) > bufferWidthMinusNewLine;
+                        if (bufferIsLargerThanPrefix && messageAndPrefixTooLargeForBuffer && _alignMessages)
+                        {
+                            // Our message may have embedded tab characters, so expand those to their space
+                            // equivalent so that wrapping works as expected.
+                            nonNullMessage = nonNullMessage.Replace("\t", consoleTab);
+
+                            // If the message and the prefix are too large for one line in the console, split the string to fit
+                            int index = 0;
+                            int messageLength = nonNullMessage.Length;
+                            // Loop until all the string has been sent to the console
+                            while (index < messageLength)
+                            {
+                                // Calculate how many chars will fit on the console buffer
+                                int amountToCopy = (messageLength - index) < (bufferWidthMinusNewLine - adjustedPrefixWidth) ? (messageLength - index) : (bufferWidthMinusNewLine - adjustedPrefixWidth);
+                                WriteBasedOnPrefix(nonNullMessage.Substring(index, amountToCopy), prefixAlreadyWritten && index == 0 && i == 0, adjustedPrefixWidth);
```

Comment:
Is the substring here passed to loggers? If not can you avoid creating it and use Span?

Reply:
In general we should aim to allocate zero strings in MSBuild, only using them when they need to go to API we can't change eg loggers, ITask, external API that require strings￼. Slicing Spans is extremely efficient.

Reply:
This is old code kept under escape hatch for compatibility reasons.